### PR TITLE
deprecate implicit default for trailing optional arguments

### PR DESCRIPTION
**Pull request by Icxolu**
This deprecates the implicit default for trailing optional (`Option<...>`) arguments. 

# Motivation
Loosely summarized from #2935, 
- The current behavior is unexpected from both Rust and Python. 
- Rust _does not_ have a concept of default values for arguments, so it's surprising that pyo3 just adds one without explicit request
- Python _does_ have default values for argument, but they have to be set explicitly.
  ```python
  def fun_1(a, b):
      ...
  def fun_2(a=None, b=None):
      ...
  
  fun_1() # <- This is errors instead of using `None` as default
  fun_2() # <- This is fine
  ```
This makes porting Python functions to Rust more surprising that it needs to be, especially if an `Option<...>` argument is _not_ in trailing position.

Expecting functions without `#[pyo3(signature=(...)]` to have a trivial signature (without defaults or other modifiers) is more intuitive. To even know about the current behavior, users have to carefully read the docs.

# Migration
This PR adds a deprecation warning to `Option<...>` arguments on functions that do _not_ currently have a
`#[pyo3(signature=(...)]` attribute and a big warning in the guide. This should push users to explicitly spell out their desired behavior during the migration period. After that we should probably make this a hard error for one release cycle, before we allow `Option<...>` arguments without `#[pyo3(signature=(...)]` again, (but without special handling).

Xref: #2935, #3735

---

~Stacked on top of #4033, so I leave this as a draft for now.~

### CODE REVIEW

**Comment by davidhewitt** [comment]
This gets a big 👍 from me! I do occasionally find the trailing optional arguments a convenient default, but just as often I've had to deliberately disable them. Having the special case also increases cognitive overhead of working out the Python signature.

Anyone got strong feelings that we shouldn't be doing this?

File: guide/src/function/signature.md [hunk 1]
```diff
@@ -121,9 +121,22 @@ num=-1
 
 ## Trailing optional arguments
 
+<div class="warning">
+
+⚠️ Warning: This behaviour is phased out 🛠️
```
Comment (davidhewitt) [inline guide/src/function/signature.md:126]:
```suggestion
⚠️ Warning: This behaviour is being phased out 🛠️
```

File: tests/ui/deprecations.stderr [hunk 2]
```diff
@@ -10,6 +10,12 @@ note: the lint level is defined here
 1  | #![deny(deprecated)]
    |         ^^^^^^^^^^
 
+error: use of deprecated function `pyo3::deprecations::deprecate_implicit_option`: Implicit default for trailing optional arguments is phased out. Add an explicit `#[pyo3(signature = (...))]` attribute a to silence this warning. In a future pyo3 version `Option<..>` arguments will be treated the same as any other argument.
+   --> tests/ui/deprecations.rs:119:39
+    |
+119 | fn pyfunction_option_2(_i: u32, _any: Option<i32>) {}
```
Comment (davidhewitt) [inline tests/ui/deprecations.stderr:16]: It would be awesome if the macro could emit the right signature here instead of `(...)`.

I wonder, what happens if we made the macro deprecation expand to something like:

```rust
#[deprecated(note = "this function has implicit defaults for the trailing `Option<T>` argument `_any`. These implicit defaults are being phased out. Add `#[pyo3(signature = (_i, _any = None))]` to this function")]
const PYFUNCTION_OPTION_2_SIGNATURE: () = ();


// immediately use the deprecated symbol to trigger the deprecation warning
const _: () = PYFUNCTION_OPTION_2_SIGNATURE;
```

I think if we attached that to the span of the whole function, rather than just an individual argument, it'd be super clear to users how to upgrade.

**Review by davidhewitt — COMMENTED** [review]
Given there doesn't seem to be much resistance, shall we move ahead with this for 0.22? I think the simplification of this edge case is a worthwhile improvement!

Just one idea regarding the deprecation message, which might be interesting to play with...

File: tests/ui/deprecations.stderr (hunk 2 above)
Comment (Icxolu) [reply tests/ui/deprecations.stderr:16]: That would be really nice! I'll give it a try

**Comment by Icxolu** [comment]
I don't really have a clue why `#![allow(deprecated)]` makes this trigger as `dead_code`. Should I just add `#[allow(dead_code)]` to the generated `const`? Or is there a better way?

File: tests/ui/deprecations.stderr [hunk 3]
```diff
@@ -10,16 +10,40 @@ note: the lint level is defined here
...
+   --> tests/ui/deprecations.rs:135:4
+    |
+135 | fn pyfunction_option_3(_i: u32, _any: Option<i32>, _foo: Option<String>) {}
+    |    ^^^^^^^^^^^^^^^^^^^
+
+error: use of deprecated constant `__pyfunction_pyfunction_option_4::SIGNATURE`: This function has implicit defaults for the trailing `Option<T>` arguments. These implicit defaults are being phased out. Add `#[pyo3(signature = (_i, _any=None, _foo=None)]` to this function to silence this warning and keep the current behavior
+   --> tests/ui/deprecations.rs:138:4
+    |
+138 | fn pyfunction_option_4(
+    |    ^^^^^^^^^^^^^^^^^^^
```
Comment (davidhewitt) [inline tests/ui/deprecations.stderr:35]: I think this is great 👍 💯

File: pyo3-macros-backend/src/deprecations.rs [hunk 4]
```diff
@@ -45,3 +48,50 @@ impl<'ctx> ToTokens for Deprecations<'ctx> {
...
 }
+
+pub(crate) fn deprecate_trailing_option_default(spec: &FnSpec<'_>) -> TokenStream {
+    if spec.signature.attribute.is_none()
+        && spec.signature.arguments.iter().any(|arg| {
+            if let FnArg::Regular(arg) = arg {
+                arg.option_wrapped_type.is_some()
+            } else {
+                false
+            }
```
Comment (davidhewitt) [inline pyo3-macros-backend/src/deprecations.rs:59]: Is any arg having `option_wrapped_type` strict enough of a check? Does it need to be something like 

1. filter to regular args
2. check if the last regular arg is an option

?

**Comment by davidhewitt** [comment]
Think it might be unrelated to the `#![allow]` and just the compiler finding the `const` isn't ending up in the binary.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=88c10b1e7f4ae498da94f3eae712178d

So I guess just add `#[allow(dead_code)]`? It is still having the effect of producing an awesome message 🚀

File: pyo3-macros-backend/src/deprecations.rs (hunk 4 above)
Comment (Icxolu) [reply pyo3-macros-backend/src/deprecations.rs:59]: I think it is enough, because we currently reject required arguments after `Option`s without a signature:
```rust
#[pyfunction]
fn foo(_i: u32, _any: Option<i32>, _foo: String) {}
```
Throws:
```
error: required arguments after an `Option<_>` argument are ambiguous
       = help: add a `#[pyo3(signature)]` annotation on this function to unambiguously specify the default values for all optional parameters
  --> src/lib.rs:82:11
   |
82 |     _foo: String,
   |           ^^^^^^
```
So if we find `Option` args without `signature` attribute they have to be trailing. Do you agree?

(But changing it to your proposal also seems relatively straight forward, so I can also do that if you prefer)

Comment (davidhewitt) [reply pyo3-macros-backend/src/deprecations.rs:59]: Ah, good point.

I wonder, does it make sense to remove that error as a follow-up to this PR? After all, in the model we're heading towards that `foo` is no longer ambiguous and is simply just 3 arguments without defaults. There's also no trailing `Option<_>` arguments which added the special case before.

That's probably a thorny enough change that I'm happy if we leave the above as-is for now and change in the follow-up...

**Review by davidhewitt — APPROVED** [review]

File: pyo3-macros-backend/src/deprecations.rs (hunk 4 above)
Comment (Icxolu) [reply pyo3-macros-backend/src/deprecations.rs:59]: > I wonder, does it make sense to remove that error as a follow-up to this PR? After all, in the model we're heading towards that `foo` is no longer ambiguous and is simply just 3 arguments without defaults.

Definitely, but I think we can only do that after the migration period, right? So If we include this in 0.22, we should require the signature attribute in 0.23 therefore can remove this ambiguity error, then in 0.24 we can remove the signature requirement again and have the new default.

Comment (davidhewitt) [reply pyo3-macros-backend/src/deprecations.rs:59]: Ah yes, of course. If we remove the error now, then users who ignore this deprecation warning and continue to use implicit trailing optional arguments may accidentally refactor them away (that's what the "ambiguity" error is designed to prevent).

**Review by davidhewitt — CHANGES_REQUESTED** [review]
Just a final thought here - we should document this in the migration guide, I think.

**Comment by Icxolu** [comment]
Good idea!

**Review by davidhewitt — APPROVED** [review]
Great!

File: guide/src/migration.md [hunk 5]
```diff
@@ -3,6 +3,18 @@
...
+## from 0.21.* to 0.22
+
+### Deprecation of implicit default for trailing optional arguments
+<details open>
+<summary><small>Click to expand</small></summary>
+
+With `pyo3` 0.22 the implicit `None` default for trailing `Option<T>` type argument is deprecated. To migrate, place a `#[pyo3(signature = (...))]` attribute on affected functions or methods and specify the desired behavior.
+The migration warning specifies the corresponding signature to keep the current behavior. With 0.23 the signature will be required for any function containing `Option<T>` type parameters to prevent accidental
+and unnoticed changes in behavior. With 0.24 this restriction will be lifted again and `Option<T>` type arguments will be treated as any other argument _without_ special handling.
+
```
Comment (davidhewitt) [inline guide/src/migration.md:15]: Where reasonable, I quite like to have before / after snippets for these blocks to give users an idea of how their code will change. Not required, but might be nice.

Otherwise, please feel free to merge 👍
